File: go/anagrams.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	if len(os.Args) != 2 {
		fmt.Println("Exactly one argument is required")
		os.Exit(1)
	}
	word := os.Args[1]
	generatePermutations(strings.Split(word, ""), len(word)-1)
}

func generatePermutations(a []string, n int) {
	if n == 0 {
		fmt.Println(strings.Join(a, ""))
	} else {
		for i := 0; i < n; i++ {
			generatePermutations(a, n-1)
			if n % 2 == 0 {
				a[0], a[n] = a[n], a[0]
			} else {
				a[i], a[n] = a[n], a[i]
			}
		}
		generatePermutations(a, n-1)
	}
}

File: js/anagrams.js
function generatePermutations(a, n) {
  if (n === 0) {
    console.log(a.join(''));
  } else {
    for (let i = 0; i < n; i++) {
      generatePermutations(a, n - 1);
      const j = n % 2 === 0 ? 0 : i;
      [a[j], a[n]] = [a[n], a[j]];
    }
    generatePermutations(a, n - 1);
  }
}

if (process.argv.length !== 3) {
  console.error('Exactly one argument is required');
  process.exit(1);
}
const word = process.argv[2];
generatePermutations(word.split(''), word.length - 1);
